text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,354 @@
+
+     ```rust
+     #[polkavm_derive::polkavm_import]
+     extern "C" {
+         fn return_ty(extension_id: u64, call_index: u32) -> u64;
+     }
+     ```
+
+     Results are SCALE-encoded bytes, with the pointer address and length packed similarly to `host_call`.
+
+- Exported Functions:
+  - `main`: The entry point of the XCQ program. It performs type checking, arguments and results passing and executes the query.
+
+### XCQ Program Execution Flow
+
+The interaction between an XCQ program and the XCQ Extension Executor follows these steps:
+
+1. Program Loading: The Executor loads the PolkaVM program binary.
+
+2. Environment Setup: The Executor configures the PolkaVM environment, registering host functions like `host_call` and `return_ty`.
+
+3. Main Function Execution: The Executor calls the program's `main` function, passing serialized query arguments.
+
+4. Program Execution:
+   1. Type Checking: The program uses the `return_ty` function to ensure compatibility with supported chain extensions.
+   2. Query Execution: The program executes the query using `host_call` and performs custom computations.
+   3. Result Serialization: The program serializes the result, writes it to shared memory, and returns the pointer and length to the executor.
+
+5. Result Retrieval: The Executor reads the result from shared memory and returns it to the caller.
+
+### XCM integration
+
+The integration of XCQ into XCM is acheived by adding a new instruction to XCM, as well as a new variant of the `Response` type in `QueryResponse` message.:
+
+- A new `ReportQuery` instruction
+
+```rust
+ReportQuery {
+  query: SizeLimitedXcq,
+  weight_limit: Option<Weight>,
+  info: QueryResponseInfo,
+}
+```
+
+Report to a given destination the results of an XCQ query. After query, a `QueryResponse` message of type `XcqResult` will be sent to the described destination.
+
+Operands:
+
+- `query: SizeLimitedXcq` has a size limit(2MB)
+  - `program: Vec<u8>`: A pre-built PVM program binary.
+  - `input: Vec<u8>`: The arguments of the program.
+- `weight_limit: WeightLimit`: The maximum weight that the query should take. `WeightLimit` is an enum that can specify either `Limit(Weight)` or `Unlimited`.
+- `info: QueryResponseInfo`: Information for making the response.
+  - `destination: Location`: The destination to which the query response message should be sent.
+  - `query_id: Compact<QueryId>`: The `query_id` field of the `QueryResponse` message
+  - `max_weight: Weight`: The `max_weight` field of the `QueryResponse` message
+
+- Add a new variant to the `Response` type in `QueryResponse`
+
+- `XcqResult = 6 (XcqResult)`
+  `XcqResult` is a enum
+  - `Ok = 0 (Vec<u8>)`: XCQ executes successfully with a SCALE-encoded response.
+  - `Err = 1 (ErrorCode)`: XCQ fails with an error code.
+`ErrorCode` is a enum
+- `ExceedMaxWeight = 0`
+- `MemoryAllocationError = 1`
+- `MemoryAccessError = 2`
+- `CallError = 3`
+- `OtherPVMError = 4`
```

Comment:
I'm wondering if there's a way to improve these errors. I guess you could run and debug your XCQ program directly on the destination chain.

Reply:
Yes everything are deterministic so replay with extra logging enabled should be the preferred way for debugging. That's how we debug XCM today.